### ISSUE TITLE
Kronecker-Factored Approximate Curvature (K-Fac) optimizer

### ISSUE THREAD

**Issue by ghost** [tensorflow/addons#2388]
**Describe the feature and the current behavior/state.**

The paper for this feature is [here](https://arxiv.org/abs/1503.05671). It's an improvement to the regular SGD algorithm used by most optimizers. There is an implementation [here](https://github.com/tensorflow/kfac) that is only tensorflow 1.x compatible and I have not found other versions that are compatible with 2.x. I need to know your suggestions / concerns / guidelines for clean outcomes. 

**Relevant information**
- Are you willing to contribute it (yes/no): yes
- Are you willing to maintain it going forward? (yes/no): yes
- Is there a relevant academic paper? (if so, where): yes, I included the link above in the description
- Is there already an implementation in another framework? (if so, where): Only a tensorflow 1.x compatible version which I included in the description
- Was it part of tf.contrib? (if so, where): no

**Which API type would this fall under (layer, metric, optimizer, etc.)** optimizer

**Who will benefit with this feature?** I'm not very informed to answer this question but being an optimizer, I think it can be as useful as commonly used optimizers: Adam, RectifiedAdam, RMSprop, ...

**Any other info.**

**Comment by james-martens** [comment]
It's great that someone is interested in doing this. I won't be able to directly contribute much myself, however I'm happy to provide advice. For example, there are extensive parts of the existing TF1 code that can be ported over with only minimal changes. (Certain other parts meanwhile will require more substantial changes.) Let me know if you'd like to meet to discuss.

**Comment by bhack** [comment]
@james-martens If it will pass the ecosystem review check can you help us to review the PR?

**Comment by james-martens** [comment]
Yeah, I can probably help. Note however that I don't know much about TF2 (or Github for that matter).

**Comment by emadboctorx** [comment]
@james-martens Sure I'm interested, how can we discuss further?

**Comment by james-martens** [comment]
Email me at jamesmartens@google.com and we can set up a video meeting.

**Comment by tomerk** [comment]
Notes from ecosystem review:
The addons optimizers seem like a reasonable home for this for now. There are no current plans to add this to tf core as far as we're aware. If it becomes widely used we could consider eventually bringing it from addons into core.

**Comment by seanpmorgan** [comment]
TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
